Projectile refactor

AmosHolland opened the pull request:
The way the effects of projectiles were handled before meant that one system had to handle all different type of projectiles. This made this system hard to understand, with constantly growing numbers of parameters.

To fix this I updated the way these collisions are handled:
- when a collision happens involving a projectile create an entity corresponding to this collision
- have a different component for all the different projectile effects
- have systems that run on entities with these components to process their effects

This has been tested and all previous projectile interactions still work.

a6-webm reviewed (approved):
Great stuff

LICHENGWEI-cloud reviewed (approved):
lgtm

AmosHolland commented:
> With this, we now have two sources of truth on what a proj does; I don't think `ProjectileHitEffect` is necessary anymore, and its variants' fields should be moved into their respective components

The reason it's like this at the moment is that I want one system to handle all of the actual collision detection, which means all projectiles should look the same when they're colliding with something, but I want different systems for each different projectile effect (to avoid a ridiculous number of parameters in the projectile effect system).

We could have a different system handle each different type of collision, which is probably the best approach. Do you have any further thoughts on this?

rosefromthedead commented:
I think that's a different question, so I'll try to explain myself better (and without misunderstanding the code). Here, flying projectiles have `ProjectileHitEffect` which describes what happens when the projectile hits. Projectiles in their colliding tick have one of `{DamageHit, BossHit}` which describes part of what effect it's going to have, but not all of it - the rest of the information is stored in the `ProjectileHitEffect`, which is a component that also applies to flying projectiles, and the type system can't stop you from applying `ProjectileHitEffect::ResetPhase` and `DamageHit` to the same entity even though that's nonsense.

I'm proposing storing a damage mask and an f32 inside `DamageEffect` so that we have a clean separation between `ProjectileHitEffect` (for flying projectiles) and `{DamageHit, BossHit}` (for colliding projectiles which will despawn next tick).

Or maybe this so that we're not duplicating field defs:
```rust
#[derive(Component)]
struct DamageHit(DamageMask, f32);

#[derive(Component)]
struct BossHit;

#[derive(Component)]
enum ProjectileHitEffect {
    Damage(DamageHit),
    ResetPhase(BossHit),
}
```

AmosHolland commented:
> I think that's a different question, so I'll try to explain myself better (and without misunderstanding the code). Here, flying projectiles have `ProjectileHitEffect` which describes what happens when the projectile hits. Projectiles in their colliding tick have one of `{DamageHit, BossHit}` which describes part of what effect it's going to have, but not all of it - the rest of the information is stored in the `ProjectileHitEffect`, which is a component that also applies to flying projectiles, and the type system can't stop you from applying `ProjectileHitEffect::ResetPhase` and `DamageHit` to the same entity even though that's nonsense.
> 
> I'm proposing storing a damage mask and an f32 inside `DamageHit` so that we have a clean separation between `ProjectileHitEffect` (for flying projectiles) and `{DamageHit, BossHit}` (for colliding projectiles which will despawn next tick).
> 
> Or maybe this so that we're not duplicating field defs:
> 
> ```rust
> #[derive(Component)]
> struct DamageHit(DamageMask, f32);
> 
> #[derive(Component)]
> struct BossHit;
> 
> #[derive(Component)]
> enum ProjectileHitEffect {
>     Damage(DamageHit),
>     ResetPhase(BossHit),
> }
> ```

Oh I see, yes, I like this change, I can make it now, but I'll need someone else to test that projectiles are still working after (if we think testing is necessary)

rosefromthedead commented:
Rebased, fixing the conflict, squashed because the history was a little messy, and removed a TODO that said to do what this PR is doing. Still not tested soz

AmosHolland commented:
> Rebased, fixing the conflict, squashed because the history was a little messy, and removed a TODO that said to do what this PR is doing. Still not tested soz

slay ty

OwenBoisse commented on src/projectile.rs:
```diff
@@ -74,75 +90,82 @@ pub fn update_linear_movement(
+}
+
+// Handle hits from boss projectiles.
+fn handle_boss_hits(
```
probs fine if everyone else is good with it but is boss_hits a bit ambigious like it could mean the boss is hit by our projectile so could be like handle_hit_by_boss idk

AmosHolland replied:
I think this is less of a problem when taken in context of the system as a whole. Like sure on this own this is ambiguous, but when the other option is `damage_hit` you can infer that it's not talking about the thing that got hit. That's just my thoughts though, would be nice to get other opinions.

AmosHolland replied:
Even in this case actually it's still not great, because `damage_hit` means the thing it does is damage, so this should really be `phase_reset_hit`.

OwenBoisse commented on src/projectile.rs:
```diff
@@ -74,75 +90,82 @@ pub fn update_linear_movement(
+                        .spawn((ProjectileHit(*e2), *t, damage_hit.clone()))
+                        .add_rollback();
+                }
+                ProjectileHitEffect::ResetPhase(boss_hit) => {
```
makes it sound like boss was hit? but maybe doesnt matter

AmosHolland commented:
> looks good just unsure on variable names

Do you know if anyone's had the chance to test the latest commits?

OwenBoisse commented:
@AmosHolland tested now and it works

OwenBoisse reviewed (approved):
lgtm
